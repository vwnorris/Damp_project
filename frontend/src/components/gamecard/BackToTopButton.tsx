import React, { useEffect, useState } from "react";
import ArrowCircleUpRoundedIcon from "@mui/icons-material/ArrowCircleUpRounded";

export const BackToTopButton = (props: {darkmode?: boolean}) => {
  const [backToTopButton, setBackToTopButton] = useState(false);

  useEffect(() => {
    window.addEventListener("scroll", () => {
      if (window.scrollY > 600) {
        setBackToTopButton(true);
      } else {
        setBackToTopButton(false);
      }
    });
  }, []);

  const scrollToTop = () => {
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
    setBackToTopButton(false);
  };

  return (
    <div data-testid="back-to-top"
      style={{
        position: "fixed",
        bottom: "20px",
        right: "50%",
      }}
    >
      {backToTopButton && (
        <ArrowCircleUpRoundedIcon
          data-testid="ArrowIcon"
          onClick={scrollToTop}
          sx={{
            backgroundColor: props.darkmode ? '#121212': 'white',
            borderRadius: "50%",
            width: 50,
            height: 50,
            '&:hover': {
              cursor: 'pointer'
            }
          }}
        />
      )}
    </div>
  );
};
