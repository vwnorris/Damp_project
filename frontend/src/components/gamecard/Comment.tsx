import { Box, Card, Divider, Rating, Stack, Typography } from '@mui/material';
import React from 'react';

export const CommentView = (props: {name: string, rating: number, comment: string}) => {

    return (
        <Card sx={{width: 1, my: 1, marginBottom:3, boxShadow:2}}>
            <Stack sx={{p:2}} gap={2}>
                <Box>
                    <Typography variant="subtitle2">User:</Typography>
                    <Typography component="div" variant="h6" style={{fontWeight: 'bold'}}>{props.name}</Typography>
                    <Divider></Divider>
                </Box>
                <Box>
                    <Typography component="legend" variant="subtitle2">Rating:</Typography>
                    <Rating name="rating" value={props.rating} readOnly/>
                </Box>
                    <Typography>{props.comment}</Typography>
            </Stack>
        </Card>
    )
}