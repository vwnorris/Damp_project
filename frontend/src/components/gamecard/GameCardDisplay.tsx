import React from "react";
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";
import CardMedia from "@mui/material/CardMedia";
import Typography from "@mui/material/Typography";
import { Box, CardActionArea, Chip, Grid } from "@mui/material";
import { observer } from "mobx-react";
import { Game } from "../../types";

// genres to color
const genreColorMap = {
  "Accounting": "#440154",
  "Action": "#470d60",
  "Adventure": "#48186a",
  "Animation & Modeling": "#482374",
  "Audio Production": "#472d7b",
  "Casual": "#453781",
  "Design & Illustration": "#424086",
  "Documentary": "#3e4989",
  "Early Access": "#3b528b",
  "Education": "#375b8d",
  "Free to Play": "#33638d",
  "Game Development": "#2f6b8e",
  "Gore": "#2c728e",
  "HTC": "#297a8e",
  "Indie": "#26828e",
  "Massively Multiplayer": "#23898e",
  "Movie": "#21918c",
  "Nudity": "#1f988b",
  "Photo Editing": "#1fa088",
  "RPG": "#22a785",
  "Racing": "#28ae80",
  "Sexual Content": "#32b67a",
  "Short": "#3fbc73",
  "Simulation": "#4ec36b",
  "Software Training": "#5ec962",
  "Sports": "#70cf57",
  "Strategy": "#84d44b",
  "Tutorial": "#98d83e",
  "Utilities": "#addc30",
  "Valve": "#c2df23",
  "Video Production": "#d8e219",
  "Violent": "#ece51b",
  "Web Publishing": "#fde725",
};

export const GameCardDisplay = observer((props: { game: Game; onClick: () => void }) => {
    
  
  const showRating = (text: string) => {
      if (text === "" || text === null) {
        return "";
      }

      const index = text.indexOf("-");

      if (text.substring(index + 2, index + 4) === "Ne") {
        return "";
      }

      return "⭐ " + text.substring(index + 2, index + 4) + "%";
    };

    const showImage = (url: string) => {
      if (url === null || url === "") {
        return (
          <CardMedia
            alt="complex"
            src="/images/default.png"
            component="img"
            height="140"
          />
        );
      }
      return <CardMedia alt="complex" src={url} component="img" height="140" />;
    };

    return (
      <Card
        sx={{
          maxWidth: 345,
          boxShadow: 3,
          margin: "auto",
          marginBottom: 1,
          marginTop: 1,
          display: "block",
          height: 1,
          backgroundColor: (theme) =>
            theme.palette.mode === "dark" ? "#1A2027" : "#fff",
        }}
        onClick={props.onClick}
        data-testid={`gameCard-${props.game.name.replace(/\s/g, '')}`}
      >
        <CardActionArea sx={{height: 1}}>
          {showImage(props.game.imagePath)}
          <CardContent sx={{height: 1}}>
            <Grid container direction="column" display="flex">
              <Grid item container direction="row" justifyContent="space-between">
                  <Grid item xs={8}>
                    <Typography
                      fontSize={"1rem"}
                      fontWeight="bold"
                    >
                      {props.game.name.replace(/(\r\n|\n|\r)/gm, "")}
                    </Typography>
                  </Grid>
                  <Grid item xs={3}>
                    <Typography variant="subtitle1" component="div">
                      {showRating(props.game.all_reviews)}
                    </Typography>
                  </Grid>
              </Grid>
              <Grid item flexGrow={1}>
                <Box
                  display="flex"
                  flexDirection="row"
                  justifyContent="space-between"
                >
                  <Typography
                    variant="body2"
                    gutterBottom
                    color="text.secondary"
                  >
                    {props.game.release_date}
                  </Typography>
                  <Typography
                    variant="body2"
                    gutterBottom
                    color="text.secondary"
                  >
                    {props.game.price === 0 ? "Free" : `${props.game.price}$`}
                  </Typography>
                </Box>
              </Grid>
              <Grid item display="flex" flexWrap="wrap" flexDirection="row">
                    {props.game.genre !== null && props.game.genre.slice(0,3).map(genre => 
                        <Chip label={genre} key={genre} sx={{mr: 0.5, mb: 0.5, backgroundColor: genreColorMap[genre as keyof typeof genreColorMap], color: 'white', fontSize: "0.8rem"}}/>
                    )}
              </Grid>
            </Grid>
          </CardContent>
        </CardActionArea>
      </Card>
    );
  }
);
