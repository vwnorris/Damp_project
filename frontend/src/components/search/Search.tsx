import {
  Box,
  Button,
  Popover,
  TextField,
  Typography,
} from "@mui/material";
import { observer } from "mobx-react";
import React, { useEffect, useRef, useState } from "react";
import HelpOutlineIcon from "@mui/icons-material/HelpOutline";
import { useStores } from "../../hooks";
import "../../App.css";
import CloseIcon from '@mui/icons-material/Close';


export const Search = observer(() => {
  
  const [openHelper, setOpenHelper] = useState<boolean>(false);
  const helperRef = useRef<HTMLDivElement>(null);

  const [openResetSearch, setOpenResetSearch] = useState<boolean>(false);
  
  const { store } = useStores();


  useEffect(() => {
    if (store.dataStore.searchString.length > 0) {
      setOpenResetSearch(true);
    } else {
      setOpenResetSearch(false);
    }
  }, [store.dataStore.searchString.length]);


  const handleOpenHelper = () => {
    setOpenHelper(true);
  };

  const handleCloseHelper = () => {
    setOpenHelper(false);
  };

  const handleClickOpen = () => {
    setOpenHelper(!openHelper);
  };


  const handleResetSearch = () => {
    store.dataStore.setSearchString("");
    store.dataStore.reloadData();
  };


  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.value !== "") {  
      setOpenResetSearch(true);
    } else {
      setOpenResetSearch(false);
    }
    store.dataStore.setSearchString(event.target.value);
    store.dataStore.reloadData();
  };

  return (
    <Box sx={{ my: 1, width: 1, display: "flex", flexDirection: "row" }}>
      <TextField
        data-testid="searchBox"
        sx={{ flexGrow: 1 }}
        value={store.dataStore.searchString}
        onChange={handleChange}
        InputProps={{
          endAdornment: (
            <div className="parent">
              
               {openResetSearch && <Button
                  data-testid="reset-search"
                  className="child"
                  sx={{ width: "fit-content", marginTop: 0}}
                  onClick={handleResetSearch}
                >
                  <CloseIcon />
                </Button>}
              
              <Typography
                className="child"
                sx={{ width: "fit-content", marginTop: 0.8  }}
                ref={helperRef}
                onClick={handleClickOpen}
                onMouseEnter={handleOpenHelper}
                onMouseLeave={handleCloseHelper}
              >
                <HelpOutlineIcon />
              </Typography>
            </div>
          ),
        }}
        label="search"
      />
      <Popover
        id="search-helper"
        sx={{ pointerEvents: "none" }}
        open={openHelper}
        anchorEl={helperRef.current}
        onClose={handleCloseHelper}
        anchorOrigin={{
          vertical: "bottom",
          horizontal: "center",
        }}
        transformOrigin={{
          vertical: "top",
          horizontal: "center",
        }}
      >
        <Typography variant="body2" sx={{ p: 2 }} onClick={handleCloseHelper}>
          You can search by game name, developer, or publisher
        </Typography>
      </Popover>
    </Box>
  );
});
