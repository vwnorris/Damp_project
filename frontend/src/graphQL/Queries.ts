import { gql } from "@apollo/client";

export const LOAD_GAMES = gql`
  query getGames($page: Int!, $search: String, $filter: FilterInput, $sort: SortInput) {
    games(page: $page, filter: $filter, sort: $sort, search: $search) {
      name
      genre
      release_date
      all_reviews
      imagePath
      publisher
      developer
      appId
      languages
      achievements
      price
      game_description
      releaseDate
    }
  }
`;

export const GET_SINGLE_GAME = gql`
  query GetGame($appId: Int!) {
    game(appId: $appId) {
      name
      genre
      release_date
      all_reviews
      imagePath
      publisher
      developer
      appId
      languages
      achievements
      price
      game_description
      comments {
        name
        rating
        comment
      }
    }
  }

`;

export const SEARCH_GAME = gql`
  query SearchGame($search: String!, $page: Int = 0, ) {
      gameSearch(search: $search, page: $page, ) {
          name
          appId
      }
  }
`;

export const ADD_COMMENT = gql`
  mutation Mutation($game: Int!, $comment: CommentInput!) {
    addComment(game: $game, comment: $comment) {
      name
      rating
      comment
    }
  }
`;
