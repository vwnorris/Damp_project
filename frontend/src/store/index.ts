import { createContext } from "react";
import { RootStore } from "./rootStore";

export const store = new RootStore()

export const defaultContext = createContext({
    store: store
});
