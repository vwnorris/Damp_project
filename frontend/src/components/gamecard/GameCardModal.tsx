import React, { useState } from 'react';
import { Accordion, AccordionDetails, AccordionSummary, Button, ButtonBase, Card, CardActions, CardContent, FormControl, Modal,  Rating, Stack, styled, TextField, Typography } from '@mui/material';
import { Comment } from '../../types';
import { CommentView } from './Comment';
import { ExpandMore } from '@mui/icons-material';
import { useMutation } from '@apollo/client';
import { ADD_COMMENT } from '../../graphQL';
import { useStores } from '../../hooks';
import { observer } from 'mobx-react';

const Img = styled('img')({
    margin: 'auto',
    display: 'block',
    maxWidth: '120%',
    maxHeight: '200%',
});

const styleCard = {
    position: 'absolute' ,
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
    width: '80%',
    height: 500,
    overflow: "hidden",
    overflowY: "scroll",
};

export const GameCardModal = observer((props: { open: boolean, onClose: ((event?: object, reason?: string) => void) | (() => void) | ((event: object, reason: string) => void) }) => {
    const [rating, setRating] = useState<number | null>(0);
    const [name, setName] = useState<string>("");
    const [comment, setComment] = useState<string>("");
    const [nameError, setNameError] = useState<boolean>(false);
    const [commentError, setCommentError] = useState<boolean>(false);
    const [addComment] = useMutation(ADD_COMMENT);

    const { store } = useStores();

    const showImage = (url: string) => {
        if (url === null || url === "") {
            return(<Img alt="complex" src="/images/default.png" />)
        }
        return(<Img alt="complex" src={url} />)
    };

    const showList = (list: string[]) => {
        if (list === null || list === undefined || list.length === 0) {
            return "..."
        }

        let textString = ""
        list.forEach(element => {
            textString += element + ", "
        });

        return textString.substring(0, textString.length - 2);
    };

    const showDescription = (text: string) => {
        let t = text.substring(300)
        let index = t.indexOf('.')

        return text.substring(17, 300 + index + 1);
        
    };

    const showPrice = (price: number) => {
        if (price === null || price === undefined || price === 0) {
            return "Free"
        }
        return "$" + price
    };

    const handleChangeRating = (_: React.SyntheticEvent, newValue: number | null) => {
        setRating(newValue);
    }
    
    const handleChangeName = (event: React.ChangeEvent<HTMLInputElement>) => {
        setName(event.target.value);
    }

    const handleChangeComment = (event: React.ChangeEvent<HTMLInputElement>) => {
        setComment(event.target.value);
    }

    const handleAddComment = async () => {
        setNameError(false);
        setCommentError(false);
        if (name === "") {
            setNameError(true);
            return;
        }

        if (comment === "") {
            setCommentError(true);
            return;
        }

        if (store.modalStore.game) {
            addComment({
                variables: {
                    game: store.modalStore.game.appId,
                    comment: {
                        name: name,
                        rating: rating,
                        comment: comment
                    }
                }
            }).then(() => {
                store.modalStore.updateGameData();
            })
            setName("")
            setComment("")
            setRating(0)
        }

    }
    // return nothing if a game is not selected
    if (store.modalStore.game === undefined) return (<></>);
    else return (
        <Modal
            open={(props.open && store.modalStore.game !== undefined)} 
            onClose={props.onClose} 
            aria-labelledby="game-modal"
            aria-describedby="modal for showing more info about a game."
            sx = {{ bgcolor: 'rgba(0, 0, 0, 0.8)' }}
        >
            <Card sx={{...styleCard}} style={{maxWidth: "650px", height: '90%' }}>
                    <ButtonBase sx={{ width: "100%", backgroundColor: "black"}}>
                        {showImage(store.modalStore.game.imagePath)}
                    </ButtonBase> 
                    <CardContent>
                        <Typography gutterBottom variant="h5" component="div">
                            {store.modalStore.game.name}
                        </Typography>
                        <Typography variant="body2" color="text.secondary">
                            {store.modalStore.game.release_date}
                        </Typography>
                        <br/>
                        <Typography gutterBottom variant="h6" component="div" >
                            Some info
                        </Typography>
                        <Typography variant="body2" color="text.secondary">
                            <span style={{fontWeight: 'bold'}}>Price: </span> {showPrice(store.modalStore.game.price)}
                        </Typography>
                        <Typography variant="body2" color="text.secondary">
                            <span style={{fontWeight: 'bold'}}>Genre: </span> {showList(store.modalStore.game.genre)}
                        </Typography>
                        <Typography variant="body2" color="text.secondary">
                            <span style={{fontWeight: 'bold'}}>Language: </span> {showList(store.modalStore.game.languages)}
                        </Typography>
                        <Typography variant="body2" color="text.secondary">
                            <span style={{fontWeight: 'bold'}}>Achievements: </span> {store.modalStore.game.achievements}
                        </Typography>
                        <br/>
                        {store.modalStore.game.game_description.length > 20 &&
                        <> 
                        <Typography gutterBottom variant="h6" component="div">
                            Description
                        </Typography>
                        <Typography variant="body2" color="text.secondary">
                            {showDescription(store.modalStore.game.game_description)}
                        </Typography> 
                        </> }
                        <br/>
                        {store.modalStore.game.all_reviews && 
                            <>  
                            <Typography gutterBottom variant="h6" component="div">
                                Reviews
                            </Typography>
                            <Typography variant="body2" color="text.secondary">
                                {store.modalStore.game.all_reviews}
                            </Typography> 
                            </>
                        }
                    </CardContent>
                    <CardActions>
                        <Accordion sx={{width: 1}}>
                            <AccordionSummary expandIcon={<ExpandMore/>} data-testid="make_comment">
                                Add comment
                            </AccordionSummary>
                            <AccordionDetails>
                                <FormControl fullWidth>
                                    <Stack gap={1}>
                                        <Stack direction='row' gap={2} sx={{width: 1, display: 'flex', alignContent: 'center' }}>
                                            <TextField label="name" placeholder='username' value={name} onChange={handleChangeName} error={nameError} required data-testid="name_textbox"/>
                                            <Stack direction='row'>
                                                <Typography component="legend">Rating</Typography>
                                                <Rating data-testid="rating" name="rating" value={rating} onChange={handleChangeRating}/>
                                            </Stack>
                                        </Stack>
                                        
                                        <TextField label="comment" placeholder='comment' value={comment} onChange={handleChangeComment} multiline rows={4} error={commentError} required data-testid="comment_textbox"/>
                                        <Button onClick={handleAddComment} sx={{ boxShadow: 2}} data-testid="add_comment">Add comment</Button>
                                    </Stack>
                                </FormControl>
                            </AccordionDetails>
                        </Accordion>
                    </CardActions>
                    <CardContent>
                        {store.modalStore.game.comments.map((value: Comment, index: number) => <CommentView key={`${value.name}-${value.rating}-${value.comment}-${index}`} name={value.name} rating={value.rating} comment={value.comment}/>)}
                    </CardContent>
                </Card>
        </Modal>
    )
});
