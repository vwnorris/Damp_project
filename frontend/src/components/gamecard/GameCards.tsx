import React, { useEffect } from "react";
import { observer } from 'mobx-react';
import { GameCardDisplay } from "./GameCardDisplay";
import { runInAction } from "mobx";
import { useStores } from '../../hooks';
import { Game } from "../../types";
import { Box, CircularProgress, Grid, Typography } from "@mui/material";
import { GameCardModal } from "./GameCardModal";
import { BackToTopButton } from './BackToTopButton';

export const GameCards = observer((props: {darkmode?: boolean}) => {
    const { store } = useStores();

    const handleScroll = (_: Event) => {
        const bottom = window.scrollY + window.innerHeight > document.body.clientHeight - 300;
        if (bottom && !store.dataStore.loading && !store.dataStore.allFound) {
            runInAction(() => {
                store.dataStore.getMoreData();
            })
        }
    }

    useEffect(() => {
        window.addEventListener('scroll', handleScroll, { passive: true });

        return () => {
            window.removeEventListener('scroll', handleScroll);
        }
    }, [])

    const handleClose = (event: object, reason: string) => {
        store.modalStore.unSelectGame();
        
    }
    
    return (
        <Grid container spacing={2} justifyContent="space-evenly" alignItems="stretch" data-testid="gameCardContainer">
            {(store.dataStore.data.length === 0 && !store.dataStore.loading) ? <Typography textAlign={'center'}>No games found :(</Typography> : <></>}

            {store.dataStore.data.map((game: Game) => 
                <Grid item key={game.appId} md={4} sm={6} xs={10}>
                    <GameCardDisplay game={game} onClick={() => store.modalStore.selectGame(game.appId)}/>
                </Grid>
            )}
            
            <GameCardModal open={store.modalStore.showModal} onClose={handleClose} />
            {store.dataStore.loading && <Box sx={{mx: 'auto', width: 'fit-content'}}><CircularProgress sx={{mx: 'auto'}}/></Box>}
            <BackToTopButton data-testid="back-to-top-button" darkmode={props.darkmode}/>
        </Grid>
    );
});
            

