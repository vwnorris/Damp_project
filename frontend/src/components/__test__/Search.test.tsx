import { Search } from "../search";
import {
  cleanup,
  fireEvent,
  render,
  screen,
  within,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockedProvider } from "@apollo/client/testing";
import { SEARCH_GAME } from "../../graphQL";
import "@testing-library/jest-dom";
import { GameCards } from "../gamecard";
import { Provider } from "mobx-react";
import { store } from "../../store";
import renderer, { act } from "react-test-renderer";

const mocks = [
  {
    request: {
      query: SEARCH_GAME,
      variables: {
        search: "trackmania",
      },
    },
    result: {
      data: {
        gameSearch: [
          {
            __typename: "Game",
            name: "Overcooked 1",
            appId: 379720,
          },
          {
            __typename: "Game",
            name: "Overcooked 2",
            appId: 578080,
          },
          {
            __typename: "Game",
            name: "Overcooked 3",
            appId: 637090,
          },
          {
            __typename: "Game",
            name: "Overcooked 4",
            appId: 221100,
          },
          {
            __typename: "Game",
            name: "Overcooked 5",
            appId: 8500,
          },
          {
            __typename: "Game",
            name: "Overcooked 5",
            appId: 601150,
          },
          {
            __typename: "Game",
            name: "Overcooked 6",
            appId: 477160,
          },
          {
            __typename: "Game",
            name: "Overcooked 7",
            appId: 644930,
          },
          {
            __typename: "Game",
            name: "Overcooked 8",
            appId: 774241,
          },
          {
            __typename: "Game",
            name: "Overcooked 9",
            appId: 527230,
          },
        ],
      },
      loading: false,
      variables: { search: "Overcooked" },
    },
  },
  {
    request: {
      query: SEARCH_GAME,
      variables: {
        search: "",
      },
    },
    result: {
      data: {
        gameSearch: [
          {
            __typename: "Game",
            name: "Overcooked 1",
            appId: 379720,
          },
          {
            __typename: "Game",
            name: "Overcooked 2",
            appId: 578080,
          },
          {
            __typename: "Game",
            name: "Overcooked 3",
            appId: 637090,
          },
          {
            __typename: "Game",
            name: "Overcooked 4",
            appId: 221100,
          },
          {
            __typename: "Game",
            name: "Overcooked 5",
            appId: 8500,
          },
          {
            __typename: "Game",
            name: "Overcooked 6",
            appId: 477160,
          },
          {
            __typename: "Game",
            name: "Overcooked 7",
            appId: 644930,
          },
          {
            __typename: "Game",
            name: "Overcooked 8",
            appId: 774241,
          },
          {
            __typename: "Game",
            name: "Overcooked 9",
            appId: 527230,
          },
        ],
      },
      loading: false,
      variables: { search: "Overcooked" },
    },
  },
];

afterEach(cleanup);

describe("Searchbar test", () => {
  it("renders with correct search", () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Search />
      </MockedProvider>
    );

    screen.getByLabelText(/search/i);
    userEvent.type(screen.getByLabelText(/search/i), "t");

    expect(store.dataStore.searchString).toBe("t");
    expect(screen.getByDisplayValue("t")).toBeInTheDocument();

    userEvent.type(screen.getByLabelText(/search/i), "{backspace}");
    expect(store.dataStore.searchString).toBe("");
  });
  
  it("help icon works as intended", () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Search />
      </MockedProvider>
    );

    screen.getByTestId("HelpOutlineIcon");
    userEvent.hover(screen.getByTestId("HelpOutlineIcon"));

    expect(
      screen.getByText("You can search by game name, developer, or publisher")
    ).toBeInTheDocument();

    userEvent.unhover(screen.getByTestId("HelpOutlineIcon"));
  });

  it("help icon removes text on click", () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Search />
      </MockedProvider>
    );
    
    userEvent.click(screen.getByTestId("HelpOutlineIcon"));
    
  })

  it("reset search", () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Search />
      </MockedProvider>
    );

    screen.getByLabelText(/search/i);
    userEvent.type(screen.getByLabelText(/search/i), "typing");

    act(() => {
      screen.getByTestId("reset-search").click();
    });

    expect(screen.queryByText("typing")).not.toBeInTheDocument();
    expect(store.dataStore.searchString).toBe("");
  })

  it("gives no games found with bad search", async () => {
    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Provider {...store}>
          <Search />
          <GameCards />
        </Provider>
      </MockedProvider>
    );
    screen.getByTestId("searchBox");
    const field = within(screen.getByTestId("searchBox")).getByLabelText(
      "search"
    );
    expect(field).toBeInTheDocument();
    fireEvent.change(field, { target: { value: "trackmania" } });
    expect(screen.getByDisplayValue("trackmania")).toBeInTheDocument();
  }); 
});


describe("Searchbar snapshot test", () => {
  it("render searcherbar", () => {
    const tree = renderer
      .create(
        <MockedProvider mocks={mocks} addTypename={false}>
          <Search />
        </MockedProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  }); 
}); 
