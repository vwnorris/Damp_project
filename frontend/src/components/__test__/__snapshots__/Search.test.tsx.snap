// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbar snapshot test render searcherbar 1`] = `
<div
  className="MuiBox-root css-77q2y4"
>
  <div
    className="MuiFormControl-root MuiTextField-root css-te7hn0-MuiFormControl-root-MuiTextField-root"
    data-testid="searchBox"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor=":r0:"
      id=":r0:-label"
    >
      search
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id=":r0:"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value="trackmania"
      />
      <div
        className="parent"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 child css-1qhucpf-MuiTypography-root"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="HelpOutlineIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
            />
          </svg>
        </p>
      </div>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-14lo706"
        >
          <span>
            search
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
